### PR TITLE
Auto-remove workflow packages from serverExternalPackages

### PR DESCRIPTION
**Pull request by TooTallNate**
## Summary

- Detect workflow-enabled packages in `nextConfig.serverExternalPackages` (`\"use step\"`, `\"use workflow\"`, or serialization patterns) and auto-remove them for the current build so Next.js can transform them.
- Keep a best-effort warning fallback in builders for non-Next `externalPackages`, and update warning text to use `externalPackages` with a Next.js mapping note.
- Update the Next API docs and changeset to reflect the new behavior.

## Why

Warning-only behavior still leaves users on a failure path at runtime. Auto-removing detected workflow packages from `serverExternalPackages` provides a safer default UX while still giving users a clear warning and manual cleanup path in `next.config`.

## Implementation details

- `withWorkflow()` now scans configured `serverExternalPackages`, filters out detected workflow packages, and passes the filtered list to both Next config and builder externals.
- The warning is emitted once per package and asks users to remove the package from `serverExternalPackages` to silence future warnings.
- Builder fallback warning remains best-effort and now:
  - works in ESM via `createRequire(import.meta.url)`
  - avoids false positives for unresolved/virtual specifiers
  - dedupes on actual warning emission

## Testing

- `pnpm vitest run packages/next/src/index.test.ts packages/builders/src/external-package-warning.test.ts`
- `pnpm build` (in `packages/next`)
- `pnpm build` (in `packages/builders`)

### CODE REVIEW

**Review by VaguelySerious — DISMISSED** [review]

File: packages/builders/src/base-builder.ts [hunk 1]
```diff
@@ -151,6 +158,85 @@ export abstract class BaseBuilder {
...
+    for (const pkg of externalPackages) {
+      if (BaseBuilder.PSEUDO_PACKAGES.has(pkg)) continue;
+      if (this.warnedExternalPackages.has(pkg)) continue;
+      this.warnedExternalPackages.add(pkg);
+
+      try {
+        // Check package.json dependencies for @workflow/serde (fast path)
+        let hasWorkflowSerdeDep = false;
+        try {
+          const pkgJsonPath = require.resolve(`${pkg}/package.json`, {
```
Comment (pranaygp) [inline packages/builders/src/base-builder.ts:291]: **Bug:** This is an ESM package (`"type": "module"` in package.json), so bare `require.resolve()` is not defined at runtime. The tests pass because vitest transpiles to CJS, but the compiled output at `dist/base-builder.js` is ESM — this will throw `ReferenceError: require is not defined` in production.

The fix already exists in this codebase — `apply-swc-transform.ts` does:
```ts
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);
```

Add the same pattern here (module-level `createRequire` + scoped `require`), or alternatively use the `enhanced-resolve` that's already imported in this file. Copilot flagged this too and I agree — this is a real bug that will silently swallow all warnings in production since both `require.resolve` calls are wrapped in try/catch.

File: packages/builders/src/base-builder.ts [hunk 2]
```diff
@@ -151,6 +158,85 @@ export abstract class BaseBuilder {
...
+
+        // Check source patterns (thorough path)
+        let hasUseStep = false;
+        let hasUseWorkflow = false;
+        let hasSerde = hasWorkflowSerdeDep;
+        try {
+          const entryPath = require.resolve(pkg, {
+            paths: [this.config.workingDir],
+          });
+          const source = await readFile(entryPath, 'utf-8');
```
Comment (pranaygp) [inline packages/builders/src/base-builder.ts:331]: **Limitation worth noting:** `require.resolve(pkg)` resolves to the package's main entry point only. If the package re-exports workflow constructs from sub-paths (e.g. `my-pkg/workflows`), or if the entry is a barrel file that re-exports from deeper modules, the regex won't see `"use step"` / `"use workflow"` in the entry file itself.

The `@workflow/serde` dep check (fast path) partially covers this for serde, but not for directives. This is probably fine as a best-effort heuristic, but worth a code comment so future maintainers understand the limitation.

File: packages/builders/src/base-builder.ts [hunk 3]
```diff
@@ -151,6 +158,85 @@ export abstract class BaseBuilder {
...
+      try {
+        // Check package.json dependencies for @workflow/serde (fast path)
+        let hasWorkflowSerdeDep = false;
+        try {
+          const pkgJsonPath = require.resolve(`${pkg}/package.json`, {
+            paths: [this.config.workingDir],
+          });
+          const pkgJsonSource = await readFile(pkgJsonPath, 'utf-8');
+          const pkgJson = JSON.parse(pkgJsonSource);
+          const deps = {
```
Comment (pranaygp) [inline packages/builders/src/base-builder.ts]: Nit: Spreading `undefined` actually works fine in JS (`{...undefined}` is `{}`), so the Copilot comment about this throwing is wrong. However, if `dependencies` is something weird like a string or array (malformed package.json), the spread would produce unexpected results. Since this is all wrapped in try/catch and best-effort, it's fine as-is — just noting for completeness.

File: packages/builders/src/base-builder.ts [hunk 4]
```diff
@@ -151,6 +158,85 @@ export abstract class BaseBuilder {
...
+   * workflow compiler when the package is externalized.
+   */
+  private async warnAboutExternalWorkflowPackages(): Promise<void> {
+    const externalPackages = this.config.externalPackages;
+    if (!externalPackages?.length) return;
+
+    for (const pkg of externalPackages) {
+      if (BaseBuilder.PSEUDO_PACKAGES.has(pkg)) continue;
+      if (this.warnedExternalPackages.has(pkg)) continue;
+      this.warnedExternalPackages.add(pkg);
```
Comment (pranaygp) [inline packages/builders/src/base-builder.ts]: Minor: The package is added to `warnedExternalPackages` before we know if a warning will actually fire. This means if detection fails (e.g. the package isn't installed yet during a partial install), the package is permanently marked as "warned" for the lifetime of this builder instance and won't be re-checked on subsequent `discoverEntries()` calls.

Since builders are typically short-lived (one build run), this is probably fine in practice. But if you want it to be more correct, move the `.add(pkg)` to just before the `console.warn()` call so only actually-warned packages are tracked.

File: packages/builders/src/external-package-warning.test.ts [hunk 5]
```diff
@@ -0,0 +1,337 @@
+import {
```
Comment (pranaygp) [inline packages/builders/src/external-package-warning.test.ts:1]: Tests are thorough and well-structured — good coverage of all detection paths, edge cases, and the dedup behavior. One thing I'd consider adding: a test that verifies the warning still fires when `package.json` resolution fails but entry-file detection succeeds (i.e., testing that the fast-path failure doesn't short-circuit the thorough path). Currently the tests assume both paths are resolvable together.

**Review by pranaygp — COMMENTED** [review]
Overall this is a well-scoped, useful addition. The detection logic is sound, tests are comprehensive, and the docs addition is clear. One real bug to fix (`require.resolve` in ESM) and a couple of minor suggestions.

**Review by karthikscale3 — COMMENTED** [review]
Here's my review of **PR #1481**: [Warn when serverExternalPackages hides workflow-enabled packages](https://github.com/vercel/workflow/pull/1481).

---

## Summary

This PR adds a **build-time warning** when packages listed in `serverExternalPackages` (or `externalPackages`) contain workflow code that would be invisible to the workflow compiler. It also adds documentation in the serialization guide explaining the externalization footgun. The problem is well-motivated -- externalized packages silently fail at runtime because the SWC transform never runs on them.

## Blocker: `require.resolve` in ESM context

Both Copilot and @pranaygp already flagged this, and I agree -- this is a **real bug** that needs fixing before merge.

`@workflow/builders` is `"type": "module"`, so bare `require.resolve()` will throw `ReferenceError: require is not defined` at runtime. The tests pass because Vitest transpiles to CJS, masking the issue. Both `require.resolve()` calls (lines 188 and 207) are wrapped in `try/catch`, so the warning logic will **silently never run** in production.

The fix is straightforward -- the codebase already has the pattern in `apply-swc-transform.ts`:

```typescript
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);
```

Add the same to `base-builder.ts` (or scope it inside the method).

## No Other Blockers

Beyond the ESM bug, the implementation is solid.

## Code Review Details

### Detection logic is well-designed

The two-tier approach is smart:
1. **Fast path**: Check `package.json` for `@workflow/serde` in deps/peerDeps
2. **Thorough path**: Read the entry file and run `detectWorkflowPatterns()` (reusing existing regex-based detection from `transform-utils.ts`)

The triple `try/catch` nesting (outer best-effort, inner fast-path, inner thorough-path) ensures the warning never blocks a build regardless of what goes wrong.

### Warning message is actionable

The warning clearly identifies: which package, what was found (step functions, workflow functions, serialization classes), and exactly how to fix it (`Remove "pkg" from serverExternalPackages`).

### Minor issues flagged in existing reviews (agree with all)

1. **`warnedExternalPackages.add(pkg)` placement** -- The package is added to the dedup set *before* checking if a warning fires. If detection fails (e.g., partial install), the package is permanently skipped for this builder instance. Since builders are short-lived, this is low-impact, but moving `.add(pkg)` to just before `console.warn()` would be more correct.

2. **Hard-coded `serverExternalPackages` in warning text** -- The config field is `externalPackages`. Non-Next.js users (Astro, SvelteKit, etc.) may be confused by the Next.js-specific name. @pranaygp's suggestion to say `externalPackages (serverExternalPackages in Next.js)` is good.

3. **Entry-point-only detection** -- `require.resolve(pkg)` only reads the main entry file. If workflow constructs are in sub-paths (e.g., `my-pkg/workflows`), regex detection misses them. The `@workflow/serde` dep check partially covers serde cases. This is acceptable as a best-effort heuristic -- worth a code comment.

### Tests are comprehensive (337 lines, 9 tests)

Covers: serde dep detection, serde symbol detection, `"use step"`, `"use workflow"`, combined patterns, pseudo-package skipping, clean packages, dedup behavior, and `@workflow/serde` in peerDependencies. One gap noted by @pranaygp: no test where `package.json` resolution fails but entry-file detection succeeds (testing that fast-path failure doesn't block thorough-path).

### Documentation addition is clear

The new "Third-Party Packages" section in `docs/content/docs/foundations/serialization.mdx` explains the problem, shows the warning message, and provides the fix. Good placement.

### Changeset is correct

Only `@workflow/builders` is listed as patch -- appropriate since the warning is contained within that package.

## CI Status

4 failures, none related:
- **Benchmark Vercel (express)**, **Benchmark Local (nitro-v3)**, **Benchmark Community World (MongoDB)** -- infra timeouts (1h+)
- **Vercel workbench-nestjs-workflow deployment** -- pre-existing NestJS deployment issue (see PR #1485)

All E2E tests pass (780 Vercel prod, 782 local dev/prod/postgres, 72 Windows). Community world failures are the usual turso/redis/mongodb pre-existing issues.

**Verdict: Fix the `require.resolve` ESM bug (use `createRequire`), then ship it. The minor suggestions (dedup timing, warning text, code comment) are nice-to-haves.**

File: docs/content/docs/foundations/serialization.mdx [hunk 6]
```diff
@@ -205,6 +205,29 @@ The `WORKFLOW_SERIALIZE` and `WORKFLOW_DESERIALIZE` methods run inside the workf
 Keep these methods simple and focused on data transformation only.
 </Callout>
 
+### Third-Party Packages
```
Comment (pranaygp) [inline docs/content/docs/foundations/serialization.mdx]: I think this warning shouldn't be in serialization docs but actually in error/troubleshooting maybe?

File: packages/builders/src/base-builder.ts [hunk 7]
```diff
@@ -151,6 +158,85 @@ export abstract class BaseBuilder {
...
+
```
[the remark this answers is not in the document]
Comment (TooTallNate) [reply packages/builders/src/base-builder.ts:286]: Fixed — `.add(pkg)` is now called just before `console.warn()`, only after detection actually succeeds.

File: packages/builders/src/base-builder.ts [hunk 8]
```diff
@@ -151,6 +158,85 @@ export abstract class BaseBuilder {
...
+        // Build a specific description of what was found
+        const issues: string[] = [];
+        if (hasUseWorkflow) issues.push('"use workflow" functions');
+        if (hasUseStep) issues.push('"use step" functions');
+        if (hasSerde) issues.push('serialization classes');
+
+        console.warn(
+          `\n${chalk.yellow('⚠')} Warning: ${chalk.bold(`"${pkg}"`)} is listed in ${chalk.bold('serverExternalPackages')} but contains workflow code (${issues.join(', ')}).` +
+            `\n  This code will ${chalk.bold('not')} be transformed by the workflow compiler, which will cause runtime failures.` +
+            `\n  Remove ${chalk.bold(`"${pkg}"`)} from ${chalk.bold('serverExternalPackages')} to fix this.\n`
```
[the remark this answers is not in the document]
Comment (TooTallNate) [reply packages/builders/src/base-builder.ts]: Fixed — warning text now says `externalPackages` (`serverExternalPackages` in Next.js).

File: packages/builders/src/base-builder.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (TooTallNate) [reply packages/builders/src/base-builder.ts:331]: Fixed — added `createRequire(import.meta.url)` at module scope.

File: packages/builders/src/base-builder.ts [hunk 9]
```diff
@@ -151,6 +158,85 @@ export abstract class BaseBuilder {
...
+        let hasWorkflowSerdeDep = false;
+        try {
+          const pkgJsonPath = require.resolve(`${pkg}/package.json`, {
+            paths: [this.config.workingDir],
+          });
+          const pkgJsonSource = await readFile(pkgJsonPath, 'utf-8');
+          const pkgJson = JSON.parse(pkgJsonSource);
+          const deps = {
+            ...pkgJson.dependencies,
+            ...pkgJson.peerDependencies,
```
[the remark this answers is not in the document]
Comment (TooTallNate) [reply packages/builders/src/base-builder.ts]: Fixed — dependencies and peerDependencies are now type-checked before use (validated as plain objects, defaulted to `{}`).

File: packages/builders/src/base-builder.ts (hunk 1 above)
Comment (TooTallNate) [reply packages/builders/src/base-builder.ts:291]: Fixed — `createRequire(import.meta.url)` is now used at module scope in base-builder.ts.

File: packages/builders/src/base-builder.ts (hunk 2 above)
Comment (TooTallNate) [reply packages/builders/src/base-builder.ts:331]: Added a code comment documenting this limitation. The primary protection is now the auto-removal in `withWorkflow()` (Next layer), so the builder warning is a best-effort fallback for non-Next integrations.

File: packages/builders/src/base-builder.ts (hunk 3 above)
Comment (TooTallNate) [reply packages/builders/src/base-builder.ts]: Agreed — spreading undefined is fine in JS. We now validate that dependencies/peerDependencies are plain objects before use anyway, to be defensive against malformed package.json.

File: packages/builders/src/base-builder.ts (hunk 4 above)
Comment (TooTallNate) [reply packages/builders/src/base-builder.ts]: Fixed — `.add(pkg)` moved to just before `console.warn()`, so only actually-warned packages are tracked.

File: packages/builders/src/external-package-warning.test.ts (hunk 5 above)
Comment (TooTallNate) [reply packages/builders/src/external-package-warning.test.ts:1]: Added — new test \"warns via entry-file detection when package.json has no serde dep\" covers the case where fast-path has no signal but entry-file detection succeeds.

File: docs/content/docs/foundations/serialization.mdx (hunk 6 above)
Comment (TooTallNate) [reply docs/content/docs/foundations/serialization.mdx]: Agreed — removed the serialization.mdx section entirely. The behavior is now documented in the `withWorkflow()` API reference instead (`docs/content/docs/api-reference/workflow-next/with-workflow.mdx`), which is more appropriate since `withWorkflow()` is what actually performs the auto-removal.

**Review by VaguelySerious — APPROVED** [review]

**Review by ijjk — APPROVED** [review]
